### PR TITLE
Add `code-snippet` component

### LINKED ISSUE #13532: UI Component - Code Snippet
This issue is to track the creation of the component: **Code Snippet**

You can access the link at https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Mobile-Components?node-id=5626%3A159428

Please have a chat with the design team before starting with the component, and don't forget to add sizing to the issue

### PR DESCRIPTION
**Pull request by andrejcesen**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Fixes #13532

### Summary

[comment]: # (Summarise the problem and how the pull request solves it)
This PR adds a `code-snippet` component that displays a block of code, including formatting and syntax highlighting. 🌈

 

### Review notes
To support syntax highlighting, this component adapts `react-syntax-highlighter` library to `react-native`. Concretely, a custom renderer was provided that maps source code's AST to `react-native` elements.

#### Platforms
- Android
- iOS

### Steps to test
Use the Preview `:quo2-snippet` screen to test it out. :)

status: ready

### CODE REVIEW

File: src/quo2/components/code/snippet.cljs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+               {:flex-shrink 1
+                ;; Round to a nearest whole number to achieve consistent
+                ;; spacing (also important for calculating `max-text-height`).
+                ;; Line height seems to be inconsistent between text being
+                ;; wrapped and text being rendered on a newline using Flexbox
+                ;; layout.
+                :line-height 18})))
+
+(defn- render-node [rows]
+  (map (fn [x]
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: Can this `x` param be name something else ?
Maybe `rendered-row`?
And can we destructure `value` and `children` here too `?

File: src/quo2/components/code/snippet.cljs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            [quo.theme :as theme]
+            [quo2.components.buttons.button :as button]
+            [quo2.components.markdown.text :as text]
+            [quo2.foundations.colors :as colors]
+            [reagent.core :as reagent]
+            [status-im.ui.components.react :as react]))
+
+;; Example themes:
+;; https://github.com/react-syntax-highlighter/react-syntax-highlighter/tree/master/src/styles/hljs
+(def themes
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: Can this be `(def ^:private themes ...)`

File: src/quo2/components/code/snippet.cljs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+           ;; Remove newlines as we already render each line separately.
+           (-> (:value x) str/trim-newline)))
+       rows))
+
+(defn- code-block [{:keys [rows line-number-width]}]
+  [into [:<>]
+   (->> rows
+        (render-node)
+        ;; Line numbers
+        (map-indexed (fn [idx x]
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: Please rename `x`

File: src/quo2/components/code/snippet.cljs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                                    ;; 18 is font's line height.
+                                    (* font-scale 18))
+            truncated? (and max-text-height (< max-text-height @text-height))
+            maybe-mask-wrapper (if truncated?
+                                 [react/masked-view
+                                  {:mask-element
+                                   (reagent/as-element
+                                    [react/linear-gradient {:colors ["black" "transparent"]
+                                                            :locations [0.75 1]
+                                                            :style {:flex 1}}])}]
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: Can we vertically align this `let` block ?

File: src/quo2/components/code/snippet.cljs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                                 [:<>])]
+
+        [rn/view {:style {:background-color background-color
+                          :border-color border-color
+                          :border-width 1
+                          :border-radius 8
+                          :overflow :hidden
+                          :padding 8
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: align maps

File: src/quo2/components/code/snippet.cljs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                          :opacity (if @text-height 1 0)}}
+         ;; Line number container
+         [rn/view {:style {:position :absolute
+                           :bottom 0
+                           :top 0
+                           :left 0
+                           :width (+ line-number-width 8 8)
+                           :background-color background-color-left
+                           :border-right-color border-color
+                           :border-right-width 1}}]
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: align this one too

File: src/quo2/components/code/snippet.cljs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+         ;; Copy button
+         [rn/view {:style {:position :absolute
+                           :bottom 8
+                           :right 8}}
+          [button/button {:icon true
+                          :type :grey
+                          :size 24
+                          :on-press on-copy-press}
+           :main-icons/copy]]]))))
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: alignment

File: src/quo2/components/code/snippet.cljs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                   :renderer (wrap-renderer-fn
+                              native-renderer {:max-lines max-lines
+                                               :on-copy-press #(when on-copy-press
+                                                                 (on-copy-press children))})
+                   ;; Default props to adapt Highlighter for react-native.
+                   :CodeTag react-native/View
+                   :PreTag react-native/View
+                   :style #js {}
+                   :customStyle #js {:backgroundColor nil}
+                   :showLineNumbers false}
```
Comment (ibrkhalil) [inline src/quo2/components/code/snippet.cljs]: Can we make this kebab cased ?

File: src/quo2/screens/code/snippet.cljs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            text (-> (:language @state) examples :text)
+            max-lines (as-> (:max-lines @state) max-lines
+                        (js/parseInt max-lines)
+                        (when-not (js/Number.isNaN max-lines)
+                          max-lines))]
+        [rn/touchable-without-feedback {:on-press rn/dismiss-keyboard!}
+         [rn/view {:padding-bottom 150}
+          [preview/customizer state descriptor]
+          [rn/view {:padding-vertical 60
+                    :padding-horizontal 16}
```
Comment (ibrkhalil) [inline src/quo2/screens/code/snippet.cljs]: Use a style prop please

**Review by ibrkhalil — CHANGES_REQUESTED** [review]
Thanks for your contribution to Status!

File: src/quo2/components/code/snippet.cljs (hunk 8 above)
Comment (andrejcesen) [reply src/quo2/components/code/snippet.cljs]: Updated all except `CodeTag` and `PreTag`, which are unfortunately not transformed by reagent. Thanks for noticing. :)

**Comment by du82** [comment]
what languages will syntax highlighting support?

**Comment by ibrkhalil** [comment]
> what languages will syntax highlighting support?

There you go

https://github.com/react-syntax-highlighter/react-syntax-highlighter/blob/master/AVAILABLE_LANGUAGES_HLJS.MD

**Comment by andrejcesen** [comment]
Sorry for accidentally removing reviewers... re-requesting a review does that apparently. :|

**Review by ibrkhalil — APPROVED** [review]
lgtm, I'll merge when we have another accepting review

**Comment by pavloburykh** [comment]
@andrejcesen thanx for the PR. Tested and ready for merge

**Review by yqrashawn — APPROVED** [review]
Thanks for your contribution. LGTM! :thumbsup:

**Pull request merged**